docs/logging-design.md

```diff
@@ -0,0 +1,51 @@
+# Design of the way AsyncHTTPClient logs
+
+<details>
+  <summary>Unless you explicitly pass AsyncHTTPClient a `Logger` instance, nothing is ever logged.</summary>
+  AsyncHTTPClient is useful in many places where you wouldn't want to log, for example a command line HTTP client. Also, we do not want to change its default behaviour in a minor release.
+</details>
+<details>
+  <summary>Nothing is logged at level `info` or higher, unless something is really wrong that cannot be communicated through the API.</summary>
+  Fundamentally, AsyncHTTPClient performs a simple task, it makes a HTTP request and communicates the outcome back via its API. In normal usage, we would not expect people to want AsyncHTTPClient to log. In certain scenarios, for example when debugging why a request takes longer than expected it may however be useful to get information about AsyncHTTPClient's connection pool. That is when enabling logging may become useful.
+</details>
+<details>
+  <summary>Each request will get a globally unique request ID (`ahc-request-id`) that will be attached (as metadata) to each log message relevant to a request.</summary>
+  When many concurrent requests are active, it can be challenging to figure out which log message is associated with which request. To facilitate this task, AsyncHTTPClient will add a metadata field `ahc-request-id` to each log message so you can first find the request ID that is causing issues and then filter only messages with that ID.
+</details>
+<details>
+  <summary>Your `Logger` metadata is preserved.</summary>
+  AsyncHTTPClient accepts a `Logger` on every request method. This means that all the metadata you have attached, will be present on log messages issued by AsyncHTTPClient.
+
+ For example, if you attach `["my-system-req-uuid": "84B453E0-0DFD-4B4B-BF22-3434812C9015"]` and then do two requests using AsyncHTTPClient, both of those requests will carry `"my-system-req-uuid` as well as AsyncHTTPClient's `ahc-request-id`. This allows you to filter all HTTP request made from one of your system's requests whilst still disambiguating the HTTP requests (they will have different `ahc-request-id`s.
+</details>
+<details>
+  <summary>Instead of accepting one `Logger` instance per `HTTPClient` instance, each request method can accept a `Logger`.</summary>
+  This allows AsyncHTTPClient to preserve your metadata and add its own metadata such as `ahc-request-id`.
+</details>
+<details>
+  <summary>Only static log messages are used.</summary>
```

Comment:
Let's call it what it is perhaps -- structured logging :-) WDYT about the below (so people can learn about the topic in general by learning about in general from various materials, rather than our swift specific static/not static terminology).


```suggestion
  <summary>All logs use the structured logging pattern, i.e. only use static log messages and accompanying key/value metadata.</summary>
```

Reply:
(+1 on the general guideline)

Reply:
@ktoso Thank you! Do you have a good link where structured logging is defined as:

- only static messages
- everything else in metadata

? Would be good to link to it here if we call it out.

(I took your change regardless)

Reply:
weirdly enough the best description I found is on the sumo logic devops glossary:

- https://www.sumologic.com/glossary/structured-logging/

> Structured logging is the practice of implementing a consistent, predetermined message format for application logs that allows them to be treated as data sets rather than text. The idea of structured logging is to take an application log that is delivered as a string of text and convert it into a simple relational data set that can be more easily searched and analyzed.

there's plenty support for the technique in most java backends, log4j2 (the most popular very good one) and niche ones; there's a good reference list here: https://tersesystems.com/#references weirdly enough they all kind of fail to write a well defined "one sentence" about it, it's become kind of known that it's "key value pairs rather than pain text" I guess.

Reply:
thanks @ktoso will link one of those!

Reply:
done